lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -185,6 +256,19 @@ def authorized_for_repo?(repo_name, user_token_is_valid, username = nil)
         false
       end
 
+      def cert_authorized_for_repo?(repo_name, uuid)
+        repository = database.connection[:repositories][{ name: repo_name }]
+        return false if repository.nil?
+        return true unless repository[:auth_required]
+
+        host = database.connection[:hosts][{ uuid: uuid }]
+        return false if host.nil?
+
+        !database.connection[:hosts_repositories].where(
+          repository_id: repository[:id], host_id: host[:id]
+        ).empty?
+      end
```

Comment:
I was wondering if this could be improved with fewer DB queries and avoid DB race conditions:

```suggestion
      def cert_authorized_for_repo?(repo_name, uuid)
        database.connection.transaction(isolation: :serializable, retry_on: [Sequel::SerializationFailure]) do
          repository = database.connection[:repositories][{ name: repo_name }]
          return false if repository.nil?
          return true unless repository[:auth_required]

          database.connection[:hosts_repositories]
            .where(repository_id: repository[:id])
            .join(:hosts, id: :host_id)
            .where(Sequel[:hosts][:uuid] => uuid)
            .any?
        end
      end
```

Reply:
I tested this and it seems to work just fine.

Reply:
After this is considered, I should be good to ACK regardless of the outcome.

Reply:
Looks like a good update..Let me test and push..

Reply:
Updated 👍🏼